File: projects/annual_report_automation/src/cli/reproduce.js
// 재현 CLI — 종이책 CSV를 읽어 세무조정을 재현하고 정답지와 대조한다.
//
//   node src/cli/reproduce.js [--dir <폴더>]
//   기본 폴더: data/templates (형식 예시). 실제 사용: data/private/fy2025 로 --dir 지정.
//
// 흐름: 입력 로드 → 무결성 검증 → 감가상각 시부인 → 조정 집계 → 별지3 계산 → 정답지 대조

import { join } from 'node:path';
import { existsSync } from 'node:fs';
import { parseFile, parseKeyValue, sum } from '../io/csv.js';
import { 검증 } from '../validate/integrity.js';
import { 시부인 } from '../engine/depreciation.js';
import { 계산 } from '../engine/tax.js';

const won = (n) => (n == null ? '—' : Math.round(n).toLocaleString('ko-KR'));

function 로드(dir) {
  const opt = (name) => existsSync(join(dir, name)) ? name : null;
  return {
    회사: parseKeyValue(join(dir, 'company.csv')),
    손익계산서: parseFile(join(dir, 'income_statement.csv')),
    재무상태표: opt('balance_sheet.csv') ? parseFile(join(dir, 'balance_sheet.csv')) : null,
    자산대장: parseFile(join(dir, 'assets.csv')),
    조정: parseFile(join(dir, 'adjustments.csv')),
    정답: opt('answer.csv') ? parseKeyValue(join(dir, 'answer.csv')) : null,
  };
}

function run(dir) {
  const { 회사, 손익계산서, 재무상태표, 자산대장, 조정, 정답 } = 로드(dir);
  console.log(`\n  재현 대상: ${dir}`);
  console.log(`  법인: 사업연도 ${회사.사업연도개시일} ~ ${회사.사업연도종료일} · ${회사.중소기업 ? '중소기업' : '일반법인'}\n`);

  // ── 1. 무결성 검증 ──────────────────────────────
  const v = 검증({ 재무상태표, 손익계산서, 자산대장 });
  console.log('  [1] 입력 무결성 검증');
  for (const c of v.checks) console.log(`      ${c.ok ? '✓' : '✗'} ${c.name} — ${c.detail}`);
  if (!v.ok) { console.log('\n  ⚠️ 입력 오류가 있습니다. 종이 원본과 대조해 수정 후 다시 실행하세요.\n'); process.exitCode = 1; return; }

  const 당기순이익 = v.당기순이익;

  // ── 2. 감가상각 시부인 (자산별) ──────────────────
  console.log('\n  [2] 감가상각 시부인');
  const 자산정규화 = 자산대장.map((a) => ({
    명: a.명, 구분: a.구분, 취득일: a.취득일,
    취득가: num(a.취득가), 기초누계: num(a.기초누계), 회사계상액: num(a.회사계상액),
    방법: a.방법, 내용연수: num(a.내용연수), 전기이월부인액: num(a.전기이월부인액),
    업무용승용차: a.업무용승용차 === 'true' || a.업무용승용차 === true,
  }));
  let 감가부인 = 0, 감가추인 = 0;
  for (const a of 자산정규화) {
    const s = 시부인(a, 회사);
    감가부인 += s.부인액; 감가추인 += s.추인액;
    const flag = s.부인액 ? `부인 +${won(s.부인액)}` : s.추인액 ? `추인 −${won(s.추인액)}` : '일치';
    console.log(`      · ${s.자산.padEnd(6)} ${s.방법}${s.내용연수}년  계상 ${won(s.회사계상액)} / 범위 ${won(s.상각범위액)}  → ${flag}`);
  }

  // ── 3. 조정 집계 (소득금액조정합계표) ────────────
  const 가산조정 = sum(조정.filter((r) => ['익금산입', '손금불산입'].includes(r.구분)), '금액') + 감가부인;
  const 차감조정 = sum(조정.filter((r) => ['손금산입', '익금불산입'].includes(r.구분)), '금액') + 감가추인;
  console.log('\n  [3] 소득금액조정합계표');
  console.log(`      가산(익금산입·손금불산입): ${won(가산조정)}  (명세서 ${won(가산조정 - 감가부인)} + 감가부인 ${won(감가부인)})`);
  console.log(`      차감(손금산입·익금불산입): ${won(차감조정)}`);

  // ── 4. 별지3 세액 계산 ──────────────────────────
  const r = 계산({
    당기순이익, 가산조정, 차감조정,
    기부금한도초과: num(회사.기부금한도초과), 이월결손금: num(회사.이월결손금),
    공제감면세액: num(회사.공제감면세액), 가산세: num(회사.가산세), 기납부세액: num(회사.기납부세액),
    중소기업: 회사.중소기업 === true, 사업연도개시일: 회사.사업연도개시일,
  });
  console.log('\n  [4] 별지3 세액조정계산서');
  for (const s of r.단계) {
    const v = s.금액 == null ? (s.주석 ?? '') : won(s.금액);
    console.log(`      ${s.부호 || ' '} ${s.라벨.padEnd(24)} ${String(v).padStart(14)}${s.최종 ? '  ★' : ''}`);
  }
  console.log(`      + 법인지방소득세 (위택스 별도)      ${won(r.지방소득세.산출세액).padStart(14)}`);
  console.log(`      = 총 납부세액                       ${won(r.총납부세액).padStart(14)}`);

  // ── 5. 정답지 대조 ──────────────────────────────
  if (정답) {
    console.log('\n  [5] 종이책 정답지 대조');
    const rows = [
      ['각사업연도소득', r.각사업연도소득, 정답.각사업연도소득],
      ['과세표준', r.과세표준, 정답.과세표준],
      ['산출세액', r.산출세액, 정답.산출세액],
      ['차감납부세액', r.차감납부세액, 정답.차감납부세액],
      ['지방소득세', r.지방소득세.산출세액, 정답.지방소득세],
    ].filter(([, , 실제]) => 실제 != null);
    let 일치 = 0;
    for (const [name, 엔진, 실제] of rows) {
      const ok = Number(엔진) === Number(실제);
      if (ok) 일치++;
      const diff = ok ? '' : `  차이 ${won(Number(엔진) - Number(실제))}`;
      console.log(`      ${ok ? '✓' : '✗'} ${name.padEnd(14)} 엔진 ${won(엔진).padStart(14)} / 실제 ${won(실제).padStart(14)}${diff}`);
    }
    const 성공 = 일치 === rows.length;
    console.log(`\n  ${성공 ? '✅ 재현 성공' : `⚠️ ${rows.length - 일치}개 불일치`} — ${일치}/${rows.length} 항목 일치`);
    if (!성공) {
      console.log('     불일치는 셋 중 하나입니다: ⓐ 입력 오타 ⓑ 엔진 로직 오류 ⓒ 세무사 판단 개입(재현 불가가 정상)');
      process.exitCode = 1;
    }
  } else {
    console.log('\n  (answer.csv 없음 — 정답지 대조 생략)');
  }
  console.log('');
}

function num(v) { return Number(String(v ?? 0).replace(/,/g, '')) || 0; }

// ── entry ──
const argv = process.argv.slice(2);
const dirIdx = argv.indexOf('--dir');
const dir = dirIdx >= 0 ? argv[dirIdx + 1] : 'data/templates';
run(dir);

File: projects/annual_report_automation/taxengine/cli/reproduce.py
"""재현 CLI — 종이책 CSV를 읽어 세무조정을 재현하고 정답지와 대조한다.

    python -m taxengine.cli.reproduce [--dir <폴더>] [--prev <전기폴더>]
    기본 폴더: data/example/fy2025 (작동 예시). 실제 사용: data/private/fy2025 로 --dir 지정.
    빈 입력 양식은 data/templates/ 에 있다 — 복사해서 채운다.
    --prev 를 주면 전기(작년) 폴더와의 연도 간 자동이월 연속성을 먼저 검증한다.

흐름: (--prev면 연속성검증 →) 무결성 검증 → 감가상각 시부인 → 조정 집계 → 별지3 → 정답지 대조
"""

import sys
from decimal import Decimal
from pathlib import Path

from taxengine.loader import parse_file
from taxengine.pipeline import 실행
from taxengine.carryover import 연속성검증
from taxengine.money import 원


def won(n) -> str:
    return "—" if n is None else f"{int(n):,}"


def run(dir_str: str, prev_str: str | None = None) -> int:
    d = Path(dir_str)
    out = 실행(d)
    회사, v, r = out["회사"], out["v"], out["r"]

    print(f"\n  재현 대상: {d}")
    print(f"  법인: 사업연도 {회사['사업연도개시일']} ~ {회사['사업연도종료일']} · "
          f"{'중소기업' if 회사.get('중소기업') else '일반법인'}")

    # 0. 연도 간 자동이월 연속성 검증 (--prev 있을 때만)
    if prev_str:
        전기 = 실행(Path(prev_str))
        c = 연속성검증(전기, out["자산대장"], 회사)
        print(f"\n  [0] 연도 간 이월 연속성 ({Path(prev_str).name} → {d.name})")
        for chk in c["checks"]:
            print(f"      {'✓' if chk['ok'] else '✗'} {chk['name']} — {chk['detail']}")
        if not c["ok"]:
            print("\n  ⚠️ 전기→당기 이월이 어긋납니다. 당기 기초값(기초누계·전기이월부인액·이월결손금)을 "
                  "전기 신고서와 대조하세요.\n")
            return 1

    # 1. 무결성 검증
    print("\n  [1] 입력 무결성 검증")
    for chk in v["checks"]:
        print(f"      {'✓' if chk['ok'] else '✗'} {chk['name']} — {chk['detail']}")
    if not v["ok"]:
        print("\n  ⚠️ 입력 오류가 있습니다. 종이 원본과 대조해 수정 후 다시 실행하세요.\n")
        return 1

    # 2. 감가상각 시부인
    print("\n  [2] 감가상각 시부인")
    for s in out["시부인들"]:
        flag = (f"부인 +{won(s['부인액'])}" if s["부인액"]
                else f"추인 −{won(s['추인액'])}" if s["추인액"] else "일치")
        print(f"      · {s['자산']:<6} {s['방법']}{s['내용연수']}년  "
              f"계상 {won(s['회사계상액'])} / 범위 {won(s['상각범위액'])}  → {flag}")

    # 3. 조정 집계
    print("\n  [3] 소득금액조정합계표")
    추진비 = out.get("추진비")
    추진비손불 = 추진비["손금불산입합계"] if 추진비 else Decimal(0)
    if 추진비:
        한도초과문구 = f", 한도초과 {won(추진비['한도초과'])}" if 추진비["한도초과"] else ""
        print(f"      · 기업업무추진비: 계상 {won(추진비['회사계상액'])} → "
              f"한도 {won(추진비['한도'])}, 증빙불비 {won(추진비['적격증빙없는금액'])}"
              f"{한도초과문구} → 손금불산입 {won(추진비손불)}")
    차량손불 = out.get("차량손금불산입", Decimal(0))
    for c in out.get("차량판정들", []):
        print(f"      · 업무용승용차: {c['사유']} → 손금불산입 {won(c['손금불산입액'])}")
    명세서분 = out["가산조정"] - out["감가부인"] - 추진비손불 - 차량손불
    추진비문구 = f" + 추진비 {won(추진비손불)}" if 추진비 else ""
    차량문구 = f" + 승용차 {won(차량손불)}" if 차량손불 else ""
    print(f"      가산(익금산입·손금불산입): {won(out['가산조정'])}  "
          f"(명세서 {won(명세서분)} + 감가부인 {won(out['감가부인'])}{추진비문구}{차량문구})")
    print(f"      차감(손금산입·익금불산입): {won(out['차감조정'])}")

    # 4. 별지3
    print("\n  [4] 별지3 세액조정계산서")
    for s in r["단계"]:
        val = s.get("주석", "") if s["금액"] is None else won(s["금액"])
        star = "  ★" if s.get("최종") else ""
        print(f"      {s['부호'] or ' '} {s['라벨']:<24} {str(val):>14}{star}")
    print(f"      + 법인지방소득세 (위택스 별도)      {won(r['지방소득세']['산출세액']):>14}")
    print(f"      = 총 납부세액                       {won(r['총납부세액']):>14}")

    # 5. 정답지 대조
    정답 = out["정답"]
    if 정답:
        print("\n  [5] 종이책 정답지 대조")
        rows = [
            ("각사업연도소득", r["각사업연도소득"], 정답.get("각사업연도소득")),
            ("과세표준", r["과세표준"], 정답.get("과세표준")),
            ("산출세액", r["산출세액"], 정답.get("산출세액")),
            ("차감납부세액", r["차감납부세액"], 정답.get("차감납부세액")),
            ("지방소득세", r["지방소득세"]["산출세액"], 정답.get("지방소득세")),
        ]
        rows = [row for row in rows if row[2] is not None]
        일치 = 0
        for name, 엔진, 실제 in rows:
            ok = 원(엔진) == 원(실제)
            일치 += ok
            diff = "" if ok else f"  차이 {won(원(엔진) - 원(실제))}"
            print(f"      {'✓' if ok else '✗'} {name:<14} "
                  f"엔진 {won(엔진):>14} / 실제 {won(실제):>14}{diff}")
        성공 = 일치 == len(rows)
        head = "✅ 재현 성공" if 성공 else f"⚠️ {len(rows) - 일치}개 불일치"
        print(f"\n  {head} — {일치}/{len(rows)} 항목 일치")
        if not 성공:
            print("     불일치는 셋 중 하나입니다: ⓐ 입력 오타 ⓑ 엔진 로직 오류 "
                  "ⓒ 세무사 판단 개입(재현 불가가 정상)")
            return 1
    else:
        print("\n  (answer.csv 없음 — 정답지 대조 생략)")
    print("")
    return 0


def main():
    argv = sys.argv[1:]

    def opt(flag):
        return argv[argv.index(flag) + 1] if flag in argv else None

    sys.exit(run(opt("--dir") or "data/example/fy2025", opt("--prev")))


if __name__ == "__main__":
    main()
